packages/connection/src/class-plugin-storage.php

```diff
@@ -0,0 +1,105 @@
+<?php
+/**
+ * Storage for plugin connection information.
+ *
+ * @package automattic/jetpack-connection
+ */
+
+namespace Automattic\Jetpack\Connection;
+
+use Automattic\Jetpack\Config;
+use WP_Error;
+
+/**
+ * The class serves a single purpose - to store the data that plugins use the connection, along with some auxiliary information.
+ * Well, we don't really store all that. The information is provided on runtime,
+ * so all we need to do is to save the data into the class property and retrieve it from there on demand.
+ *
+ * @todo Adapt for multisite installations.
+ */
+class Plugin_Storage {
+
+	/**
+	 * Connected plugins.
+	 *
+	 * @var array
+	 */
+	private static $plugins = array();
+
+	/**
+	 * Whether the plugins were configured.
+	 * To make sure we don't call the configuration process again and again.
+	 *
+	 * @var bool
+	 */
+	private static $plugins_configuration_finished = false;
+
+	/**
+	 * Add or update the plugin information in the storage.
+	 *
+	 * @param string $slug Plugin slug.
+	 * @param array  $args Plugin arguments, optional.
+	 *
+	 * @return bool
+	 */
+	public static function upsert( $slug, array $args = array() ) {
+		self::$plugins[ $slug ] = $args;
```

Comment:
Do we want to verify for duplicated slugs or we'll just let it override whatever was registered first?

Reply:
That's a really good point.
I was planning to leave it for the future versions as not so pressing at the moment, but if you have time, may be you could look into it yoursef.

The problem is that the connection package can be initialized multiple times by the same plugin, so we can't simply check if it's already been already initialized for that slug.
My first thought was to prevent it from adding its info into the plugin list, but unless we through a fatal error there (we shouldn't do that) that doesn't really change anything.

The plugins with a duplicated slug will use the same connection basically, and both disconnect if one of them gets disconnected - the same thing that happens now.

So it appears to me that the best we can do is the following:
1. In the `Plugin::add()` we will check if the slug has already been added.
2. If it has, we throw `_doing_it_wrong()`, informing the developer that they shouldn't do that.
3. As an option, we could also add the `Plugin::update()` method and refer developers to it if they actually want to update the plugin info in runtime, and don't run the `Plugin_Storage::upsert()` in the `add()` method.

That's quite a lousy protection, because it's very unlikely to be caught during development, but that's the only one I have. That's why I've been hesitant about implementing it and decided to put it off. But now we have more muscle 💪 in the connection team, so it would be great if you could look into the problem and see if we can come up with a better solution.

Reply:
> the connection package can be initialized multiple times by the same plugin

What exactly do you mean by "initialize the connection package" ?

(Sure, I can help with that now!)

Reply:
I mean creating the object: `$manager = new Manager( 'plugin-slug' )`.
Since it can be done multiple times, we can't really check for duplicates there.